cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+# libcudf C++ Developer Guide
+
+This document serves as a guide for contributors to libcudf C++ code. Developers should also refer 
+to these additional files for further documentation of libcudf best practices.
+
+* [Documentation Guide](DOCUMENTATION.md) for guidelines on documenting libcudf code.
+* [Testing Guide](TESTING.md) for guidelines on writing unit tests.
+* [Benchmarking Guide](TODO) for guidelines on writing unit benchmarks.
+
+# Overview
+
+libcudf is a C++ library that provides GPU-accelerated data-parallel algorithms for processing 
+column-oriented tabular data. libcudf provides algorithms including slicing, filtering, sorting, 
+various types of aggregations, and database-type operations such as grouping and joins. libcudf
+serves a number of clients via multiple language interfaces, including Python and Java. Users may
+also use libcudf directly from C++ code.
+
+## Lexicon
+
+This section defines terminology used within libcudf
+
+### Column
+
+A column is an array of data of a single type. Along with Tables, columns are the fundamental data 
+structures used in libcudf. Most libcudf algorithms operate on columns. Columns may have a validity
+mask representing whether each element is valid or null (invalid). Columns of nested types are 
+supported, meaning that a column may have child columns.
+
+### Element
+
+An individual data item within a column. Also known as a row.
+
+### Scalar
+
+A type representing a single element of a data type.
+
+### Table
+
+A table is a collection of columns. It is also known as a dataframe.
+
+### View
+
+A view is a non-owning object that provides zero-copy access (possibly with slicing or offsets) data 
```

Comment:
I think a section on slicing and column_view offsets is warranted.  It's an easy-to-overlook case and the word "offset" is heavily overloaded in cudf.

Reply:
Added